src/main/java/org/stellar/sdk/Sep10Challenge.java

```diff
@@ -40,4 +46,366 @@ public static String newChallenge(
+   * Verifies that for a SEP-0010 challenge transaction
+   * all signatures on the transaction are accounted for and that the signatures
+   * meet a threshold on an account. A transaction is verified if it is signed by
+   * the server account, and all other signatures match a signer that has been
+   * provided as an argument, and those signatures meet a threshold on the
+   * account.
+   *
+   * @param challengeTransaction SEP-0010 transaction challenge transaction in base64.
+   * @param serverAccountId      Account ID for server's account.
+   * @param network              The network to connect to for verifying and retrieving.
+   * @param threshold            The threshold on the client account.
+   * @param signers              The signers of client account.
+   * @return a list of signers that were found is returned, excluding the server account ID.
+   * @throws InvalidSep10ChallengeException If the SEP-0010 validation fails, the exception will be thrown.
+   * @throws IOException                    If read XDR string fails, the exception will be thrown.
+   */
+  public static List<String> verifyChallengeTransactionThreshold(String challengeTransaction, String serverAccountId, Network network, int threshold, List<Signer> signers) throws InvalidSep10ChallengeException, IOException {
+    String ed25519PublicKeySignerType = "ed25519_public_key";
+
+    if (signers == null || signers.isEmpty()) {
+      throw new InvalidSep10ChallengeException("No signers provided.");
+    }
+
+    List<String> ed25519PublicKeySigners = new ArrayList<String>();
+    for (Signer signer : signers) {
+      if (ed25519PublicKeySignerType.equals(signer.getType())) {
+        ed25519PublicKeySigners.add(signer.getKey());
+      }
+    }
+
+    List<String> signersFound = verifyChallengeTransactionSigners(challengeTransaction, serverAccountId, network, ed25519PublicKeySigners);
+
+    int weight = 0;
+    for (String signer : signersFound) {
+      for (Signer signerWithWeight : signers) {
+        if (signer.equals(signerWithWeight.getKey()) && ed25519PublicKeySignerType.equals(signerWithWeight.getType())) {
+          weight += signerWithWeight.getWeight();
+          break;
+        }
+      }
+    }
+
+    if (weight < threshold) {
+      throw new InvalidSep10ChallengeException(String.format("Signers with weight %d do not meet threshold %d.", weight, threshold));
+    }
+
+    return signersFound;
+  }
+
+  private static List<String> verifyTransactionSignatures(Transaction transaction, List<String> signers) throws InvalidSep10ChallengeException {
+    List<DecoratedSignature> signatures = transaction.getSignatures();
+    if (signatures.isEmpty()) {
+      throw new InvalidSep10ChallengeException("Transaction has no signatures.");
+    }
+    List<String> signersFound = new ArrayList<String>();
+    for (String signer : signers) {
+      if (verifyTransactionSignedByAccountId(transaction, signer)) {
+        signersFound.add(signer);
+      }
+    }
```

Comment:
This logic verifies each signer has signed the transaction independently, but it would be more defensive to ensure that all signatures are matched to a signer in the list exactly once. Without doing that this code is making the assumption that whatever keys are in use will always work in such a way that no signature could be matched to more than one key and might one day result in two signers being matched to a single signature and a signature being double counted.

@overcat I raised this on the Python SDK also here https://github.com/StellarCN/py-stellar-base/issues/252.

It would be better if we can use the same logic as is used in the Go SDK for signature verification:
https://github.com/stellar/go/blob/a135d590c455964bf1bfdbb33ba266b5a08f9e53/txnbuild/transaction.go#L652-L695